kube/src/client/discovery.rs

```diff
@@ -0,0 +1,341 @@
+    }
+
+    /// Returns served versions (e.g. `["v1", "v2beta1"]`) of this group.
+    /// This list is always non-empty, and sorted in the following order:
+    /// - Stable versions (with the last being the first)
+    /// - Beta versions (with the last being the first)
+    /// - Alpha versions (with the last being the first)
+    /// - Other versions, alphabetically
+    // Order is documented here:
+    // https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions
+    pub fn versions(&self) -> impl Iterator<Item = &str> {
+        let versions = self.versions_and_resources.as_slice();
+        versions.iter().map(|ver_data| ver_data.version.as_str())
+    }
+
+    /// Returns preferred version for working with given group.
+    /// This function tries to use `preferredVersion` field at first,
+    /// and then uses some fallbacks if that is missing.
+    pub fn preferred_version(&self) -> &str {
+        match &self.preferred_version {
+            Some(v) => v,
+            None => self.versions().next().unwrap(),
+        }
+    }
+
+    /// Returns resources available in version `ver` of this group.
+    /// If the group does not support this version,
+    /// returns empty vector.
+    pub fn resources_by_version(&self, ver: &str) -> Vec<GroupVersionKind> {
+        let resources = self
+            .versions_and_resources
+            .iter()
+            .find(|ver_data| ver_data.version == ver)
+            .map(|ver_data| ver_data.resources.as_slice())
+            .unwrap_or(&[]);
+        resources
+            .iter()
+            .cloned()
+            .map(|mut api_resource| {
+                api_resource.group = Some(if self.name == "core" {
+                    String::new()
+                } else {
+                    self.name.clone()
+                });
+                api_resource.version = Some(ver.to_string());
+                GroupVersionKind::from_api_resource(&api_resource, "unused/v0")
+            })
+            .collect()
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+enum Version {
+    Stable(u32),
+    Beta(u32, Option<u32>),
+    Alpha(u32, Option<u32>),
+    Nonconformant(String),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+enum VersionSortKey<'a> {
+    Stable(Reverse<u32>),
+    Beta(Reverse<u32>, Reverse<Option<u32>>),
+    Alpha(Reverse<u32>, Reverse<Option<u32>>),
+    Nonconformant(&'a str),
+}
+
+impl Version {
+    fn to_sort_key(&self) -> VersionSortKey {
+        match self {
+            Version::Stable(v) => VersionSortKey::Stable(Reverse(*v)),
+            Version::Beta(v, beta) => VersionSortKey::Beta(Reverse(*v), Reverse(*beta)),
+            Version::Alpha(v, alpha) => VersionSortKey::Alpha(Reverse(*v), Reverse(*alpha)),
+            Version::Nonconformant(nc) => VersionSortKey::Nonconformant(nc),
+        }
+    }
+
+    fn try_parse(v: &str) -> Option<Version> {
```

Comment:
>  should be probably be a FromStr impl to match rust conventions

I don't have a strong opinion here since this is a private type. Given you think it'd better, I'll use `FromStr`.

> does anything return a Nonconformant version

We discover not only built-in APIs but also `APIService`s and `CustomResourceDefinition`s. AFAIK their versions can be arbitrary, and we have to support it.

Reply:
Ah. I see. That makes sense. Maybe a comment about why it's there would be helpful here.

I was mainly advocating for FromStr on the expectation that there was an error case we should just bail at, but if it's for crds and people can return whatever in crds then this setup is fine.